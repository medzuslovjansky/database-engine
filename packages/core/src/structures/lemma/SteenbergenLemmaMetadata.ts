export type SteenbergenLemmaMetadata = {
  id: number;
  addition?: string;
  partOfSpeech: string;
  type: number;
  sameInLanguages?: string;
  genesis?: string;
  frequency?: number;
  using_example?: string;
};
