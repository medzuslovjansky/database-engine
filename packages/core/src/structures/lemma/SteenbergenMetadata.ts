export type SteenbergenMetadata = {
  id: number;
  addition?: string;
  partOfSpeech: string;
  type: number;
  sameInLanguages?: string;
  genesis?: string;
  frequency?: number;
  using_example?: string;
};
