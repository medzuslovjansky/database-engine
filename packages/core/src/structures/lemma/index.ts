export * from './Lemma';
export * from './InterslavicLemma';
