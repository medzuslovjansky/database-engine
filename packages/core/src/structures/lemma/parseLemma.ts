export function parseLemma(rawStr: string) {
  const str = trim(rawStr);
  const leftN = str.lastIndexOf('(');
  const rightN = str.lastIndexOf(')');
  if (rightN < leftN || (leftN === -1 && rightN !== -1)) {
    throw new Error(`Lemma value has incorrect parentheses: ${str}`);
  }

  let value: string;
  let annotations: string[];

  if (rightN === str.length - 1) {
    value = str.slice(0, leftN).trimEnd();
    annotations = str
      .slice(leftN + 1, rightN)
      .split(';')
      // eslint-disable-next-line unicorn/no-array-callback-reference
      .map(trim);
  } else {
    value = str;
    annotations = [];
  }

  if (annotations[0] === '') {
    annotations.splice(0, 1);
  }

  return { value, annotations };
}

function trim(str: string): string {
  return str.trim();
}
