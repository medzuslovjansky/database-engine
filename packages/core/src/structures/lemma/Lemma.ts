import type { SteenbergenLemmaMetadata } from './SteenbergenLemmaMetadata';
import { parseLemma } from './parseLemma';

export type LemmaOptions = {
  value: string;
  annotations: string[];
};

export class Lemma {
  constructor(options: Partial<LemmaOptions> = {}) {
    this.value = options.value ?? '';
    this.annotations = options.annotations ?? [];
  }

  public value: string;

  public annotations: string[];

  public steen?: SteenbergenLemmaMetadata;

  public clone(): Lemma {
    return new Lemma({
      value: this.value,
      annotations: [...this.annotations],
    });
  }

  public toString(): string {
    if (this.annotations.length === 0) {
      return this.value;
    }

    return `${this.value} (${this.annotations.map(String).join('; ')})`;
  }

  public static parse(str: string): Lemma {
    return new Lemma(parseLemma(str));
  }
}
