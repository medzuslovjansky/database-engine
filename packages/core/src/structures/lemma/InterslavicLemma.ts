import { Lemma } from './Lemma';

export class InterslavicLemma extends Lemma {
  public steen?: SteenbergenLemmaMetadata;

  public clone(): InterslavicLemma {
    const result = super.clone() as InterslavicLemma;
    if (this.steen) {
      result.steen = { ...this.steen };
    }

    return result;
  }
}

export type SteenbergenLemmaMetadata = {
  id: number;
  addition?: string;
  partOfSpeech: string;
  type?: number;
  sameInLanguages?: string;
  genesis?: string;
  frequency?: number;
  using_example?: string;
};
