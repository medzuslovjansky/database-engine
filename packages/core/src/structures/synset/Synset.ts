import { isIterable } from '../../utils';
import { Lemma } from '../lemma';

import { parseSynset } from './parseSynset';

export type SynsetOptions = SynsetMetadata & {
  lemmas: Lemma[];
};

export type SynsetMetadata = {
  verified?: boolean;
  debatable?: boolean;
};

type EqualityPredicate<T> = (a: T, b: T) => boolean;

export class Synset {
  public verified: boolean;
  public debatable: boolean;
  public readonly lemmas: Lemma[];

  constructor(options: Partial<SynsetOptions> = {}) {
    this.verified = options.verified ?? false;
    this.debatable = options.debatable ?? false;
    this.lemmas = options.lemmas ?? [];
  }

  public clone(): Synset {
    return new Synset({
      verified: this.verified,
      debatable: this.debatable,
      lemmas: this.lemmas.map((g) => g.clone()),
    });
  }

  public isEmpty(): boolean {
    return this.lemmas.length === 0;
  }

  public get size(): number {
    return this.lemmas.length;
  }

  public add(value: Lemma | string | Iterable<Lemma> | Iterable<string>): this {
    if (isIterable(value) && typeof value !== 'string') {
      for (const val of value) {
        this.add(val);
      }

      return this;
    }

    let lemma: Lemma;

    if (typeof value === 'string') {
      lemma = Lemma.parse(value);
    } else if (value instanceof Lemma) {
      lemma = value.clone();
    } else {
      throw new TypeError('Invalid value type: ' + value);
    }

    this.lemmas.push(lemma);

    return this;
  }

  public intersection(
    other: Synset,
    equals: EqualityPredicate<Lemma> = valueEquals,
  ): Synset {
    const result = new Synset({
      verified: this.verified && other.verified,
      debatable: this.debatable || other.debatable,
    });

    for (const l1 of this.lemmas) {
      for (const l2 of other.lemmas) {
        if (equals(l1, l2)) {
          result.add(l1.clone());
          break;
        }
      }
    }

    return result;
  }

  public union(
    other: Synset,
    equals: EqualityPredicate<Lemma> = valueEquals,
  ): Synset {
    const result = new Synset({
      verified: this.verified && other.verified,
      debatable: this.debatable || other.debatable,
    }).add(this.lemmas);

    for (const l2 of other.lemmas) {
      let exists = false;
      for (const l1 of this.lemmas) {
        if (equals(l1, l2)) {
          exists = true;
          break;
        }
      }

      if (!exists) {
        result.add(l2.value);
      }
    }

    return result;
  }

  public difference(
    other: Synset,
    equals: EqualityPredicate<Lemma> = valueEquals,
  ): Synset {
    const result = new Synset({
      verified: this.verified && other.verified,
      debatable: this.debatable || other.debatable,
    });

    for (const l1 of this.lemmas) {
      let exists = false;
      for (const l2 of other.lemmas) {
        if (equals(l1, l2)) {
          exists = true;
          break;
        }
      }

      if (!exists) {
        result.add(l1.clone());
      }
    }

    return result;
  }

  public clear(): this {
    this.lemmas.splice(0, Number.POSITIVE_INFINITY);
    return this;
  }

  public toString(): string {
    const hasCommas = this.lemmas.some((l) => l.value.includes(','));

    return (
      (this.debatable ? '#' : '') +
      (this.verified ? '' : '!') +
      this.lemmas.map(String).join(hasCommas ? '; ' : ', ')
    );
  }

  public static parse(str: string) {
    return new Synset(parseSynset(str));
  }
}

function valueEquals(a: Lemma, b: Lemma): boolean {
  return a.value === b.value;
}
