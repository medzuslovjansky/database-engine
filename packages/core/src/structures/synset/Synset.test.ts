import { Lemma } from '../lemma';

import { Synset } from './Synset';

describe('Synset', () => {
  describe('when created', () => {
    describe('with no options', () => {
      it('should contain no lemmas', () => {
        expect(anEmptySynset().lemmas).toEqual([]);
      });

      it('should not report itself as verified', () => {
        expect(anEmptySynset().meta.verified).toBe(undefined);
      });

      it('should not report itself as debatable', () => {
        expect(anEmptySynset().meta.debatable).toBe(undefined);
      });

      it('should be empty', () => {
        expect(anEmptySynset().isEmpty()).toBe(true);
      });
    });

    describe('with options', () => {
      it('should contain those lemmas', () => {
        const { synset, lemmas } = aComplexSynset();
        expect(synset.lemmas).toEqual(lemmas);
      });

      it('should contain meta.verified', () => {
        const { synset, meta } = aComplexSynset();
        expect(synset.meta.verified).toBe(meta.verified);
      });

      it('should contain meta.debatable', () => {
        const { synset, meta } = aComplexSynset();
        expect(synset.meta.debatable).toBe(meta.debatable);
      });

      it('should be not empty', () => {
        const { synset } = aComplexSynset();
        expect(synset.isEmpty()).toBe(false);
      });
    });
  });

  describe('.add()', () => {
    it('should add a string', () => {
      const synset = anEmptySynset();
      expect(synset.add('value').toString()).toBe('!value');
    });

    it('should add a string array', () => {
      const synset = anEmptySynset();
      expect(synset.add(['value1', 'value2']).toString()).toBe(
        '!value1, value2',
      );
      expect(synset.lemmas.length).toBe(2);
    });

    it('should add a string array (with commas)', () => {
      const synset = anEmptySynset();
      expect(synset.add(['value1, value2', 'value3']).toString()).toBe(
        '!value1, value2; value3',
      );
      expect(synset.lemmas.length).toBe(2);
    });

    it('should add a lemma', () => {
      const synset = anEmptySynset();
      const lemma = new Lemma({
        value: 'value',
        annotations: ['some annotation'],
      });

      expect(synset.add(lemma).toString()).toBe('!value (some annotation)');
    });

    it('should add an array of lemmas', () => {
      const synset = anEmptySynset();
      const lemma1 = new Lemma({ value: 'value', annotations: ['annot.'] });
      const lemma2 = new Lemma('value2');

      expect(synset.add([lemma1]).add([lemma2]).toString()).toBe(
        '!value (annot.), value2',
      );
    });

    it('should throw an attempt to pass something else', () => {
      const synset = anEmptySynset();
      const smth = {} as any;
      expect(() => synset.add(smth)).toThrowError(/Cannot add/);
    });
  });

  describe('.clear()', () => {
    it('should mutate and clear lemmas', () => {
      const synset = anEmptySynset().add('something');
      synset.clear();
      expect(synset.toString()).toBe('!');
    });
  });

  describe('.clone()', () => {
    it('should deep clone the synset', () => {
      const synset = anEmptySynset().add(
        new Lemma({
          value: 'hello',
          annotations: ['greeting'],
        }),
      );

      const clone = synset.clone();

      expect(clone.toString()).toBe('!hello (greeting)');
      expect(clone.toString()).toBe(synset.toString());
      expect(clone).not.toBe(synset);
      expect(clone.lemmas).not.toBe(synset.lemmas);
      expect(clone.lemmas[0]).not.toBe(synset.lemmas[0]);
      expect(clone.lemmas[0].annotations).not.toBe(
        synset.lemmas[0].annotations,
      );
    });
  });

  describe('.size', () => {
    it('should return lemma count', () => {
      const synset = anEmptySynset();
      expect(synset.size).toBe(0);
      synset.add(['some1', 'some2']);
      expect(synset.size).toBe(2);
    });
  });

  describe('.union()', () => {
    it('should create a new synset with all the values', () => {
      const synset1 = anEmptySynset().add(['a', 'b']);
      const synset2 = anEmptySynset().add(['b', 'c']);
      const union = synset1.union(synset2);

      expect(union).not.toBe(synset1);
      expect(union).not.toBe(synset2);
      expect(union.toString()).toBe('!a, b, c');
    });

    it('should always take the worst-case metadata', () => {
      const synset1 = anEmptySynset().add(['a', 'b']);
      const synset2 = anEmptySynset();
      synset1.meta.verified = false;
      synset1.meta.debatable = true;

      const union = synset1.union(synset2);
      expect(union.toString()).toBe('#!a, b');
    });
  });

  describe('.difference()', () => {
    it('should create a new synset with values missing in the other one', () => {
      const synset1 = anEmptySynset().add(['a', 'b']);
      const synset2 = anEmptySynset().add(['b', 'c']);
      const difference = synset1.difference(synset2);

      expect(difference).not.toBe(synset1);
      expect(difference).not.toBe(synset2);
      expect(difference.toString()).toBe('!a');
    });

    it('should always take the worst-case metadata', () => {
      const synset1 = anEmptySynset().add(['a', 'b']);
      const synset2 = anEmptySynset();
      synset1.meta.verified = false;
      synset1.meta.debatable = true;

      const union = synset1.difference(synset2);
      expect(union.toString()).toBe('#!a, b');
    });
  });

  describe('.intersection()', () => {
    it('should return an empty synset if there is no intersection', () => {
      const s1 = anEmptySynset().add('lemma1');
      const s2 = anEmptySynset().add('lemma2');

      expect(s1.intersection(s2).toString()).toBe('!');
    });

    it('should return a non-empty synset if there is an intersection', () => {
      const s1 = anEmptySynset().add(['this', 'lemma']);
      const s2 = anEmptySynset().add(['that', 'lemma']);

      s1.meta.verified = s2.meta.verified = true;
      s2.meta.debatable = true;

      expect(s1.intersection(s2).toString()).toBe('#lemma');
    });
  });

  describe('.lemmas()', () => {
    it('should iterate over lemmas across lemmas', () => {
      const { lemmas, synset } = aComplexSynset();
      expect(synset.lemmas).toEqual(lemmas);
    });
  });

  describe('when stringified', () => {
    it('should prepend ! if it is not verified', () => {
      const { synset } = aComplexSynset();
      synset.meta.verified = false;
      synset.meta.debatable = false;

      expect(`${synset}`).toMatch(/^!/);
    });

    it('should not prepend ! if it is verified', () => {
      const { synset } = aComplexSynset();
      synset.meta.verified = true;
      synset.meta.debatable = false;

      expect(`${synset}`).not.toMatch(/^!/);
    });

    it('should prepend # if it is debatable', () => {
      const { synset } = aComplexSynset();
      synset.meta.debatable = true;
      synset.meta.verified = true;

      expect(`${synset}`).toMatch(/^#[^!]/);
    });

    it('should not prepend # if it is not debatable', () => {
      const { synset } = aComplexSynset();
      synset.meta.debatable = false;
      synset.meta.verified = false;

      expect(`${synset}`).not.toMatch(/^#/);
    });

    it('should prepend both #! if it is verified and debatable', () => {
      const { synset } = aComplexSynset();
      synset.meta.verified = false;
      synset.meta.debatable = true;

      expect(`${synset}`).toMatch(/^#!/);
    });

    it('should separate lemmas with (;) and lemmas with (,) inside', () => {
      const synset = anEmptySynset();
      synset.add('toj, ktory rabi');
      synset.add('rabotajuci');
      synset.lemmas[0].annotations.push('adj.');

      expect(`${synset}`).toBe('!toj, ktory rabi (adj.); rabotajuci');
    });
  });

  function aComplexSynset() {
    const lemma1 = new Lemma({ value: 'you' });
    const lemma2 = new Lemma({ value: 'thou', annotations: ['obsolete'] });
    const lemma3 = new Lemma({ value: 'lupus', annotations: ['medical'] });
    const lemmas = [lemma1, lemma2, lemma3];

    const verified = true;
    const debatable = true;

    return {
      synset: new Synset({ lemmas, debatable, verified }),
      meta: {
        verified,
        debatable,
      },
      lemmas,
      annotations: ['obsolete', 'medical'],
    };
  }

  function anEmptySynset() {
    return new Synset();
  }
});
