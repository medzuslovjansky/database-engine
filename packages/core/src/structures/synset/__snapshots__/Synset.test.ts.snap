// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Synset #parse() should parse synset: !: synset 1`] = `
Synset {
  "lemmas": [],
  "verified": false,
}
`;

exports[`Synset #parse() should parse synset: !U-Boot: synset 1`] = `
Synset {
  "lemmas": [
    Lemma {
      "annotations": [],
      "value": "U-Boot",
    },
  ],
  "verified": false,
}
`;

exports[`Synset #parse() should parse synset: !only, ĝuste nun: synset 1`] = `
Synset {
  "lemmas": [
    Lemma {
      "annotations": [],
      "value": "only",
    },
    Lemma {
      "annotations": [],
      "value": "ĝuste nun",
    },
  ],
  "verified": false,
}
`;

exports[`Synset #parse() should parse synset: : synset 1`] = `
Synset {
  "lemmas": [],
  "verified": true,
}
`;

exports[`Synset #parse() should parse synset: bring (up) to: synset 1`] = `
Synset {
  "lemmas": [
    Lemma {
      "annotations": [],
      "value": "bring (up) to",
    },
  ],
  "verified": true,
}
`;

exports[`Synset #parse() should parse synset: extra comma,: synset 1`] = `
Synset {
  "lemmas": [
    Lemma {
      "annotations": [],
      "value": "extra comma",
    },
  ],
  "verified": true,
}
`;

exports[`Synset #parse() should parse synset: extra,,comma,: synset 1`] = `
Synset {
  "lemmas": [
    Lemma {
      "annotations": [],
      "value": "extra",
    },
    Lemma {
      "annotations": [],
      "value": "comma",
    },
  ],
  "verified": true,
}
`;

exports[`Synset #parse() should parse synset: з'явитися: synset 1`] = `
Synset {
  "lemmas": [
    Lemma {
      "annotations": [],
      "value": "з'явитися",
    },
  ],
  "verified": true,
}
`;

exports[`Synset #parse() should parse synset: за (напр.: по грибы, за хлебом), по: synset 1`] = `
Synset {
  "lemmas": [
    Lemma {
      "annotations": [
        "напр.: по грибы, за хлебом",
      ],
      "value": "за",
    },
    Lemma {
      "annotations": [],
      "value": "по",
    },
  ],
  "verified": true,
}
`;

exports[`Synset #parse() should parse synset: и; а зато (смысл2): synset 1`] = `
Synset {
  "lemmas": [
    Lemma {
      "annotations": [],
      "value": "и",
    },
    Lemma {
      "annotations": [
        "смысл2",
      ],
      "value": "а зато",
    },
  ],
  "verified": true,
}
`;

exports[`Synset #parse() should parse synset: сей (устар.; местоим.): synset 1`] = `
Synset {
  "lemmas": [
    Lemma {
      "annotations": [
        "устар.",
        "местоим.",
      ],
      "value": "сей",
    },
  ],
  "verified": true,
}
`;

exports[`Synset #parse() should parse synset: той, що є;: synset 1`] = `
Synset {
  "lemmas": [
    Lemma {
      "annotations": [],
      "value": "той, що є",
    },
  ],
  "verified": true,
}
`;
